### ISSUE TITLE
startsWith can't infer string literal types

### ISSUE THREAD

**Issue by omonk** [gustavoguichard/string-ts#181]
```typescript
import { startsWith } from "string-ts";

const a = {
  id: "id-1" as `id-${string}`
};

const b = startsWith(a.id, "id", 0);
// b = boolean

const c = startsWith("id-1231", "id", 0);
// c = true
```

I'm unsure if this is pushing at the limits of TS/string-ts but this would be awesome to have available!

**Comment by gustavoguichard** [comment]
Hey @omonk ! It seems feasible and a good suggestion!
I'll take a look if I can solve it or perhaps some of our contributors @p9f @jly36963 @hverlin @mjuksel

**Comment by p9f** [comment]
I think it is a direct consequence of https://github.com/gustavoguichard/string-ts/pull/105 - not sure how we can have the best of both world here, agree that would be great though

**Comment by p9f** [comment]
(@omonk I took the liberty to edit your comment to remove the `name` part as I think it was a mistake - please change it back if it was by purpose)

**Comment by omonk** [comment]
@p9f - ah yes, wasn't vital for displaying the example!

**Comment by jly36963** [comment]
I'll take a look after work today

**Comment by p9f** [comment]
I tried 

```typescript
export type StartsWith<
  T extends string,
  S extends string,
  P extends number = 0,
> = All<[IsStringLiteral<S>, IsNumberLiteral<P>]> extends true
  ? Math.IsNegative<P> extends false
    ? P extends 0
      ? T extends `${S}${string}`
        ? true
        : `${S}${string}` & T extends never
        ? false
        : boolean
      : StartsWith<Slice<T, P>, S, 0> // P is >0, slice
    : StartsWith<T, S, 0> // P is negative, ignore it
  : boolean
```

which is _slightly_ better than what we have currently as it passes the following additional tests:
```ts
  type test7 = Expect<Equal<StartsWith<`abc${string}`, 'a'>, true>>
  type test8 = Expect<Equal<StartsWith<`a${string}`, 'a', number>, boolean>>
```

but it still fails on 
```ts
  type test9 = Expect<Equal<StartsWith<`b${string}`, 'a'>, false>>
```

because, for some reason, `a${string}` & `m${string}` does not extends `never`....

**Comment by mjuksel** [comment]
Could try this one? @jly36963 @p9f @gustavoguichard 

```ts
export type StartsWith<
  T extends string,
  S extends string,
  P extends number = 0,
  C extends 0[] = []
> = All<[IsStringLiteral<T | S>, IsNumberLiteral<P>]> extends true
  ? Math.IsNegative<P> extends true
    ? false
    : P extends 0
    ? T extends `${S}${string}`
      ? true
      : false
    : C['length'] extends P
    ? T extends `${S}${string}`
      ? true
      : false
    : StartsWith<Slice<T, 1>, S, P, [...C, 0]>
  : false;
```

**Comment by gustavoguichard** [comment]
Thank y'all for the suggestions. @jly36963 has put on a nice PR that solves it.
@omonk I just released a minor version with this fix: string-ts@2.1.0

**Comment by omonk** [comment]
Thanks so much everyone! Will take it out for a sping
